# Context

**This dataset will be used to enhance book recommendations for Goodreads users.**

In [1]:
import pandas as pd

In [14]:
from pathlib import Path

def load_raw_data(filename):
    
    project_dir = Path.cwd().parent
    raw_data_dir = project_dir / "data" / "raw"
    
    file_path = raw_data_dir / filename
    
    if not file_path.exists():
        raise FileNotFoundError(f"Dataset not found: {file_path}")
    
    print(f"Loading dataset from: {file_path}")
    
    return pd.read_csv(file_path)

In [15]:
books = load_raw_data("goodreads_books.csv")
books.head()

Loading dataset from: f:\VScode\ai-apprenticeship-session6\ai-apprenticeship-02-data-lifecycle\group1\data\raw\goodreads_books.csv


,id,title,link,series,cover_link,author,author_link,rating_count,review_count,average_rating,...,isbn13,asin,settings,characters,awards,amazon_redirect_link,worldcat_redirect_link,recommended_books,books_in_series,description
0,630104,Inner Circle,https://www.goodreads.com//book/show/630104.In...,(Private #5),https://i.gr-assets.com/images/S/compressed.ph...,"Kate Brian, Julian Peploe",https://www.goodreads.com/author/show/94091.Ka...,7597,196,4.03,...,9781416950417,NaN,NaN,NaN,NaN,https://www.goodreads.com//book_link/follow/17...,https://www.goodreads.com//book_link/follow/8?...,"726458, 726458, 1537534, 3047848, 1651302, 304...","381489, 381501, 352428, 630103, 1783281, 17832...",Reed Brennan arrived at Easton Academy expecti...
1,9487,A Time to Embrace,https://www.goodreads.com//book/show/9487.A_Ti...,(Timeless Love #2),https://i.gr-assets.com/images/S/compressed.ph...,Karen Kingsbury,https://www.goodreads.com/author/show/3159984....,4179,177,4.35,...,9781595542328,NaN,NaN,NaN,NaN,https://www.goodreads.com//book_link/follow/17...,https://www.goodreads.com//book_link/follow/8?...,"127352, 127352, 40642197, 127353, 127354, 3891...",115036,"Ideje az ölelésnek Történet a reményről,..."
2,6050894,Take Two,https://www.goodreads.com//book/show/6050894-t...,(Above the Line #2),https://i.gr-assets.com/images/S/compressed.ph...,Karen Kingsbury,https://www.goodreads.com/author/show/3159984....,6288,218,4.23,...,NaN,NaN,"Bloomington, Indiana(United States)",NaN,NaN,https://www.goodreads.com//book_link/follow/17...,https://www.goodreads.com//book_link/follow/8?...,"706250, 706250, 666481, 11942636, 706241, 1273...","4010795, 40792877, 7306261",Filmmakers Chase Ryan and Keith Ellison have c...
3,39030,Reliquary,https://www.goodreads.com//book/show/39030.Rel...,(Pendergast #2),https://i.gr-assets.com/images/S/compressed.ph...,"Douglas Preston, Lincoln Child",https://www.goodreads.com/author/show/12577.Do...,38382,1424,4.01,...,9780765354952,NaN,"New York State(United States) New York City, N...","Aloysius X.L. Pendergast, Margo Green, William...",NaN,https://www.goodreads.com//book_link/follow/17...,https://www.goodreads.com//book_link/follow/8?...,"39026, 39026, 11007668, 3986318, 51887330, 194...","67035, 39031, 39033, 136637, 136638, 30068, 39...",NaN
4,998,The Millionaire Next Door: The Surprising Secr...,https://www.goodreads.com//book/show/998.The_M...,NaN,https://i.gr-assets.com/images/S/compressed.ph...,"Thomas J. Stanley, William D. Danko",https://www.goodreads.com/author/show/659.Thom...,72168,3217,4.04,...,9780671015206,NaN,NaN,NaN,Independent Publisher Book Award (IPPY) Nomine...,https://www.goodreads.com//book_link/follow/17...,https://www.goodreads.com//book_link/follow/8?...,"1052, 1052, 763362, 69571, 78427, 30186948, 10...",NaN,The incredible national bestseller that is cha...


In [4]:
books.shape

(52199, 31)

In [5]:
books.columns

Index(['id', 'title', 'link', 'series', 'cover_link', 'author', 'author_link',
       'rating_count', 'review_count', 'average_rating', 'five_star_ratings',
       'four_star_ratings', 'three_star_ratings', 'two_star_ratings',
       'one_star_ratings', 'number_of_pages', 'date_published', 'publisher',
       'original_title', 'genre_and_votes', 'isbn', 'isbn13', 'asin',
       'settings', 'characters', 'awards', 'amazon_redirect_link',
       'worldcat_redirect_link', 'recommended_books', 'books_in_series',
       'description'],
      dtype='object')

In [6]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 52199 entries, 0 to 52198
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   id                      52199 non-null  int64  
 1   title                   52199 non-null  object 
 2   link                    52199 non-null  object 
 3   series                  23319 non-null  object 
 4   cover_link              51592 non-null  object 
 5   author                  52199 non-null  object 
 6   author_link             52199 non-null  object 
 7   rating_count            52199 non-null  int64  
 8   review_count            52199 non-null  int64  
 9   average_rating          52199 non-null  float64
 10  five_star_ratings       52199 non-null  int64  
 11  four_star_ratings       52199 non-null  int64  
 12  three_star_ratings      52199 non-null  int64  
 13  two_star_ratings        52199 non-null  int64  
 14  one_star_ratings        52199 non-null

In [7]:
books.isna().sum().sort_values(ascending=False)

asin                      46963
awards                    41573
settings                  40684
characters                38504
books_in_series           30121
series                    28880
original_title            12949
isbn13                    12692
isbn                      11883
recommended_books          3994
worldcat_redirect_link     3985
publisher                  3636
genre_and_votes            2840
description                2575
number_of_pages            2330
date_published              860
cover_link                  607
title                         0
one_star_ratings              0
two_star_ratings              0
three_star_ratings            0
four_star_ratings             0
five_star_ratings             0
average_rating                0
review_count                  0
rating_count                  0
author_link                   0
amazon_redirect_link          0
author                        0
link                          0
id                            0
dtype: i

In [8]:
books.describe()

,id,rating_count,review_count,average_rating,five_star_ratings,four_star_ratings,three_star_ratings,two_star_ratings,one_star_ratings,number_of_pages
count,5.219900e+04,5.219900e+04,52199.000000,52199.000000,5.219900e+04,5.219900e+04,52199.000000,52199.000000,52199.000000,49869.000000
mean,1.015443e+07,1.887361e+04,1012.980881,4.020610,7.817176e+03,6.250785e+03,3456.512424,935.496600,413.640051,328.942730
std,1.164443e+07,1.163978e+05,4054.802421,0.367161,5.876373e+04,3.473533e+04,18249.298631,5890.077215,3843.356500,252.790119
min,1.000000e+00,0.000000e+00,0.000000,0.000000,0.000000e+00,0.000000e+00,0.000000,0.000000,0.000000,0.000000
25%,3.628190e+05,3.400000e+02,31.000000,3.820000,1.170000e+02,1.090000e+02,63.000000,15.000000,6.000000,210.000000
50%,6.383537e+06,2.295000e+03,163.000000,4.030000,8.100000e+02,7.650000e+02,452.000000,107.000000,36.000000,304.000000
75%,1.738060e+07,9.297500e+03,622.000000,4.230000,3.375500e+03,3.190500e+03,1866.000000,450.000000,151.000000,392.000000
max,5.444299e+07,6.801077e+06,169511.000000,5.000000,4.414877e+06,1.868421e+06,980183.000000,529060.000000,537793.000000,14777.000000


In [9]:
books.duplicated().sum()

0

In [10]:
books.nunique().sort_values()

average_rating              262
number_of_pages            1367
one_star_ratings           3118
two_star_ratings           4904
asin                       5236
settings                   5338
review_count               5627
date_published             9120
awards                     9157
three_star_ratings         9597
publisher                 11058
characters                12276
four_star_ratings         12399
five_star_ratings         12808
rating_count              19267
books_in_series           20318
series                    22654
author_link               23009
author                    27737
original_title            37816
isbn13                    39507
isbn                      40316
genre_and_votes           47301
recommended_books         48047
worldcat_redirect_link    48214
description               49252
title                     49685
cover_link                51592
amazon_redirect_link      52199
link                      52199
id                        52199
dtype: i

In [11]:
books[['average_rating','rating_count','review_count']].corr()

,average_rating,rating_count,review_count
average_rating,1.000000,0.020095,0.009104
rating_count,0.020095,1.000000,0.832072
review_count,0.009104,0.832072,1.000000


# Data Exploration

**1. Initial exploration**

- You can use the DataFrame methods from pandas, such as DataFrame.head(), DataFrame.info(), DataFrame.describe(), etc, to investigate the data.
- Alternatively use the following package:

In [12]:
from ydata_profiling import ProfileReport

profile = ProfileReport(cust, title="Retail Dataset Profiling")

profile.to_file("retail_profile_report.html")

NameError: name 'cust' is not defined

The dataset contains 52,199 book records with 31 columns, representing a mixture of book metadata, user engagement metrics, and web reference links.

Several categories of information are present:

    Book Metadata

        title

        author

        publisher

        number_of_pages

        date_published

        series

        books_in_series

    User Engagement Metrics

        rating_count

        review_count

        average_rating

        five_star_ratings

        four_star_ratings

        three_star_ratings

        two_star_ratings

        one_star_ratings

Identifiers

        id

        isbn

        isbn13

        asin

External Links

        link

        cover_link

        amazon_redirect_link

        worldcat_redirect_link

The dataset contains a combination of numeric features (ratings and counts) and text-based metadata (authors, titles, descriptions).

Initial exploration also revealed that several metadata columns contain large numbers of missing values, particularly:

    asin

    awards

    settings

    characters

    books_in_series

    series

The dataset contains no duplicate rows, which indicates that each entry represents a unique book record.

- What are your initial observations about the data?

**2. Connect to the business context**
   
Based on your exploration, which columns would be most helpful in building a book recommendation system?

The most useful columns for building a book recommendation system are those representing reader engagement and book characteristics.

Key Engagement Metrics

    average_rating

    rating_count

    review_count

    five_star_ratings

    four_star_ratings

    three_star_ratings

    two_star_ratings

    one_star_ratings

These features describe how readers interact with and rate books. They provide useful signals for identifying highly rated or popular books.

A correlation analysis also showed a strong relationship between rating_count and review_count, indicating that books with many ratings typically also receive many reviews.

Useful Book Metadata

    title

    author

    genre_and_votes

    number_of_pages

    date_published

These fields provide contextual information about the book and could support content-based recommendation methods.

Columns Likely to Be Excluded

Several columns are unlikely to be useful for modelling because they represent identifiers or links rather than meaningful features:

    id

    link

    cover_link

    amazon_redirect_link

    worldcat_redirect_link

**3. Identify the data quality issues**

Identify Data Quality Issues in your selected columns. Now that you've chosen the columns you believe are most relevant, perform a detailed analysis to identify specific issues like missing values, inconsistent formats, incorrect data types, etc,  within those columns only.

High Levels of Missing Data

Some metadata columns contain very large numbers of missing values. Examples include:

    asin

    awards

    settings

    characters

    books_in_series

    series

In many cases these columns are missing values for a large proportion of the dataset, suggesting that the information is optional or inconsistently collected.

Sparse Metadata Fields

Columns such as characters, settings, and awards contain textual information but are only populated for a subset of books. This limits their usefulness without additional processing.

Mixed Feature Types

The dataset contains a mixture of:

    numeric features (ratings and counts)

    categorical text features

    identifiers

    web links

Some fields such as genre_and_votes may require additional parsing or transformation before they can be used in machine learning models.

Irrelevant Columns for Modelling ?

Several columns represent external links or identifiers, which are not useful predictive features and would likely be removed during the data preparation stage or can be used to obtain data from linked sources but brings in issues around useage of the data and its only quality issues.

**4. Initial data cleaning and transformation**

(Each person from the group)Choose a relevant column with one or multiple data issues and try to address them. For example: missing values, incorrect data type, mixed types of values, etc.

In [21]:
books['genre_and_votes'].isna().sum()

np.int64(2840)

In [22]:
books['genre_and_votes'].isna().mean()

np.float64(0.0544071725511983)

In [23]:
books['genre_and_votes'].dropna().head(10)

0              Young Adult 161, Mystery 45, Romance 32
1    Christian Fiction 114, Christian 45, Fiction 3...
2      Christian Fiction 174, Christian 81, Fiction 58
3    Thriller 626, Mystery 493, Horror 432, Fiction...
4    Economics-Finance 1162, Nonfiction 910, Busine...
5    Romance 550, Historical-Historical Fiction 316...
6    Romance 725, Historical-Historical Fiction 395...
7    Fiction 106, American-Southern 26, Gothic-Sout...
8    Sequential Art-Comics 947, Sequential Art-Grap...
9    Poetry 599, Childrens 384, Classics 301, Ficti...
Name: genre_and_votes, dtype: str

In [24]:
books['genre_and_votes'].nunique()

47301

In [25]:
books['genre_and_votes'].dropna().str.len().describe()

count    49359.000000
mean        86.181466
std         48.086315
min          5.000000
25%         54.000000
50%         78.000000
75%        107.000000
max        280.000000
Name: genre_and_votes, dtype: float64

In [26]:
books['genre_and_votes'].dropna().str.extract(r'(\d+)').head()

,0
0,161
1,114
2,174
3,626
4,1162


The genre_and_votes column contains valuable recommendation information, but it is currently stored as unstructured text.

The main issues are:

    Multiple values stored in a single column

    Genre names and vote counts combined in text format

    High uniqueness preventing direct categorical modelling

To make the data usable, the column should be transformed by:

    splitting genres into separate values

    extracting vote counts

    creating structured features such as:

        primary genre

        genre vote totals

        binary genre indicators

This transformation would allow the genre information to be used effectively in a recommendation model.

**5. Bonus task: Early feature engineering**

Discuss and implement the potential feature engineering opportunities in this dataset. What new columns can you create from the existing ones that might enhance the next steps of the AI project?

In [27]:
books['genre_count'] = books['genre_and_votes'].dropna().str.split(',').str.len()

In [28]:
books['positive_rating_ratio'] = (
    books['four_star_ratings'] + books['five_star_ratings']
) / books['rating_count']

In [42]:
from pathlib import Path

project_dir = Path.cwd().parent
interim_dir = project_dir / "data" / "interim"

interim_dir.mkdir(exist_ok=True)

interim_file = interim_dir / "goodreads_books_interim.csv"

books.to_csv(interim_file, index=False)

print("Saved interim dataset to:", interim_file)

Saved interim dataset to: f:\VScode\ai-apprenticeship-session6\ai-apprenticeship-02-data-lifecycle\group1\data\interim\goodreads_books_interim.csv


Two new features were created from the existing dataset to improve the potentially improve the usefulness of the data for future modelling tasks.

1. Genre Count

The genre_and_votes column contains multiple genres stored in a comma-separated text format. To better represent this information, a new column called genre_count was created to capture how many genres are associated with each book.

books['genre_count'] = books['genre_and_votes'].dropna().str.split(',').str.len()

This feature indicates the number of genres assigned to a book. Books associated with multiple genres may appeal to a wider audience and could influence recommendation patterns.

2. Positive Rating Ratio
A new feature called positive_rating_ratio was created using the star rating distribution columns.

books['positive_rating_ratio'] = (
    books['four_star_ratings'] + books['five_star_ratings']
) / books['rating_count']

This feature represents the proportion of highly positive ratings relative to the total number of ratings.

Compared with raw counts, this ratio provides a clearer signal of how positively readers perceive a book. It helps distinguish between books that are widely rated and books that are consistently rated highly.

Why These Features Are Useful

These engineered features provide more meaningful information than the original raw fields.

Feature	Benefit
genre_count	captures how broadly a book spans different genres
positive_rating_ratio	measures reader sentiment more clearly

These help convert raw data into structured signals that are more suitable for machine learning models, particularly recommendation systems.